src/main/java/org/javarosa/core/model/User.java

```diff
@@ -88,7 +93,15 @@ public void writeExternal(DataOutputStream out) throws IOException {
     }
 
     public String getUsername() {
-        return username;
+        if (!isPlatformKeyStoreAvailable()) {
+            return this.username;
+        } else {
+            try {
+                return EncryptionUtils.decryptUsingKeyFromKeyStore(this.username, USER_CREDENTIALS_KEY_ALIAS);
```

Comment:
I think this will cause issues for transitions from old CC version to new CC version. If `User` is already store in DB decrypted, we will try to decrypt a non-encrypted user and fail.

Reply:
@shubham1g5 I don't think this is the case, I also had this concern and I ran a few tests around it and worked fine. Something to keep in mind here is that we are only encrypting the credentials when bringing them to memory, this shouldn't affect the way we save/retrieve the credentials to/from the DB. During the login process, we use the credentials provided by the `user` (which are in `plaintext`) to fetch and verify the user, the `username` is stored in `plaintext` and we don't decrypt the `pwd`, we use the one provided by the user to set the cached pwd, right? Am I missing something?

Reply:
We do[ store the fields of this class in SQl DB](https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/commcare/core/parse/UserXmlParser.java#L99) in user db.  I am worried that if we load the user class from the previously stored row in DB and call `getUsername` on it, we will be in a bad state due to calling `decryptUsingKeyFromKeyStore` on decrypted user name.

Reply:
I think to test it you will need to run the apk without these changes, login with a user, install apk with these changes, login (should be a local login without using internet), Go to a place in code that calls getUserName on this class and test the behaviour.

Reply:
@shubham1g5 I ran the test as suggested and no issues there too, I also did the opposite and didn't have any problems. The code highlighted above stores the parsed user to the database after a sync, again this shouldn't be affected by this change. The areas we should be concerned are (1) saving the user directly from memory (without using `getUsername` or `getCachePwd`) (2) Retrieving the user from the database and loading it to memory without using any of the constructors or setters. About retrieving from the database, we actually [do it](https://github.com/dimagi/commcare-core/blob/52746e8b30fafa44aa3da2ac1648d1426be498ef/src/main/java/org/commcare/core/parse/UserXmlParser.java#L51C28-L51C28) but before storing, but I think this is fine.

Reply:
@shubham1g5 just to provide more clarity, this is how the credentials are now stored in memory:
![image](https://github.com/dimagi/commcare-core/assets/19228119/6fae6b17-9ba3-4409-97d9-2938a2580b0c)
and here's how it looks with this change
![image](https://github.com/dimagi/commcare-core/assets/19228119/3a9de89c-3fd3-496e-95dd-d1ccd8c185f3)
the password hash remains the same.

Reply:
I still think it's a very dangerous change that it depends on someone to know right order of calls on a Java POJO class. The user class is meant to be stored in DB and we can't expect the caller of the class to not call save on the class after calling a getter method of the class. As such we should expect that any changes to `User` fields can be saved in DB and consequently loaded and we should not make any changes to it that we don't want to get saved. 

Probably the right solution here is to add another wrapper class called `EncryptedUser` which has same fields as `User` model but stores them encrypted in memory. And then while loading the `User` fields we always use `EncryptedUser` and not `User` model. 

@ctsims Also tagging you to see if you agree on the right course of action here.

Reply:
@shubham1g5 To me the code looks like it fulfills the case you are describing? I agree that the cycle of 

load -> read  fields-> write

shouldn't change the underlying data. As written it seems like the pattern is that data is stored encrypted but when the write method is called, the data is decrypted again on the way back into storage. IE: This only affects the in-memory object as long as only the protected (getUsername / setUsername) methods are used exclusively. Is that right @avazirna ?

My only concern is actually overloading the `username` field itself to serve both purposes potentially causing coding issues. I'd probably make separate fields for the encrypted slug and a plaintext password to clarify the internal `getUsername()` / `setUsername()` code. 

Would it be worth adding some unit tests that specifically address the read/write cycle that Shubham is describing and showing that disabling a 'mock' key store in between doesn't prevent the username from being read. (demonstrating that it's not getting written as an encrypted value)

Reply:
>  data is stored encrypted but when the write method is called, the data is decrypted again on the way back into storage.

Maybe I am missing something but If someone writes a code like - 
````
User u = getUser()
String username = u.getUsername()
write(storage, u)
````
Would not this end up writing encrypted username in the storage without one explicitly making any edits ?

> when the write method is called, the data is decrypted again on the way back into storage

I think this is what I am not seeing in the current code. 

>  I'd probably make separate fields for the encrypted slug

Do you mean have 2 fields  like `username` and `encryptedUserName` ?


+1 for unit tests.

Reply:
Ahh nevermind, I was not paying attention to change in `writeExternal`. I see how the data will end up getting decrypted before write now. 

Think it would be more clear to have separate methods for encryption rather than using getters and setters. 
````
{
   private String username;
   
   readExternal() {
      setUsername(encrypt(ExtUtil.readString(in)));
   }
   
   writeExternal() {
     ExtUtil.writeString(out, decrypt(getUsername()));
   }
   
   getUsername(){
     return username;
   }
   
   setUsername(username){
     username = username;
   }

}
````

Reply:
I think the fact that all of us interpreted the code a little differently implies that it would help do adjust the structure a bit, but I differ a bit on the preferred refactor, as I think in the stated approach `username` is ambiguously encrypted in a way that isn't clear from the class internals.

That was why my proposal was to have to separate class members (`encryptedUsername` and `plaintextUsername`) which I think would make the code structure clearer since it eliminates the concept that one of the internal storage variables is cannonical. I think that would make it unambigious from the outside whether `getUsername()` and `setUsername()` are operating in plaintext mode, but maybe I'm overestimating.

Reply:
@ctsims having `plaintextUsername` would defeat the reason we need the encrypted version, the goal is to avoid storing the user credentials in memory in plaintext. One option I initially entertained was to have a custom annotation, let's say @EncryptInMemory, at the field level that would be used to generate the getters and setters during build. Any thoughts on that?

Reply:
> Think it would be more clear to have separate methods for encryption rather than using getters and setters.

True, but this would mean `encrypting`/`decrypting` before each `set`/`get` call, right?

Reply:
> This only affects the in-memory object as long as only the protected (getUsername / setUsername) methods are used exclusively. Is that right @avazirna ?

Correct, @ctsims.

Reply:
> Probably the right solution here is to add another wrapper class called `EncryptedUser` which has same fields as `User` model but stores them encrypted in memory. And then while loading the `User` fields we always use `EncryptedUser` and not `User` model.

I could explore this, @shubham1g5. My question is, wouldn't wrapping the `User` imply loading it into memory?

Reply:
@avazirna - The reason I mentioned `plaintextUsername` is because currently the code as written still does allow for that username to be used if there's no locally registered encryption provider in the platform. Right now there's basically "soft forked" behavior where the meaning of the `username` field is overloaded depending on how the platform is using it. My proposed change is just to respect this split in behavior but to make it more explicit at the code level, and to prevent, say, issues that could arise from the meaning of the `username` field changing at runtime (IE: Right now if the field is set when a provider isn't available, and then one is registered, the app would attempt a decrypt read on the ambiguous `username` field which has a plaintext value.)

Reply:
@ctsims @shubham1g5 I ended up splitting the fields as suggested (1069a52b934259569219cf466a20d5f1ad4f6cdf) and also added a unit test ([8c74193](https://github.com/dimagi/commcare-android/pull/2723/commits/8c741930c9291db96d2453f3cdcb5335d0868a47)) and mocks for the Android key store and key generator ([7bfd5af](https://github.com/dimagi/commcare-android/pull/2723/commits/7bfd5af673304bf2bc3986b90cd0b3187b26170c))